File: C++/isPail.cpp
#include <iostream>
#include <vector>
#include <algorithm>

struct ListNode {
    int val;
    struct ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution{
    public:
     bool isPail(ListNode* head) {
        // write code here
        std::vector<int>v;
        ListNode*current =head;
        while(current){
            v.push_back(current->val);
            current =current->next;
        }
        return std::equal(v.begin(),v.end(),v.rbegin());
    }
};

int main(){
    std::vector<int>v={1,2,2,1};
    Solution solution;

    ListNode* head=new ListNode(v[0]);
    ListNode* current =head;
    for(int i =1;i<v.size();i++){
        current->next=new ListNode(v[i]);
        current =current->next;
    }

    bool result =solution.isPail(head);
    if(result){
        std::cout<<"The linked list is palindrome."<<std::endl;
    }
    else{
        std::cout<<"The linked list is not a palindrome."<<std::endl;
    }

    current=head;
    while(current){
        ListNode* next=current->next;
        delete current;
        current=next;
    }
    return 0;
}

File: python/isPail.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def isPail(self , head: ListNode) -> bool:
        list1 = []
        while head:
            list1.append(head.val)
            head = head.next
        for i in range(len(list1)):
            if i>len(list1)-1-i:
                break
            if list1[i] != list1[len(list1)-1-i]:
                return False
        return True

head=ListNode(1)
head.next=ListNode(2)
head.next.next=ListNode(2)
head.next.next.next=ListNode(1)
solution=Solution()
while head:
    print(head.val,end='')
    head=head.next
pail_head=solution.isPail(head)
print(end=":")
if(pail_head):
    print("true")
else:
    print("false")